### DISCUSSION TITLE
Captcha in carelink

### DISCUSSION THREAD

**Discussion by lmbigdata** [NightscoutFoundation/xDrip#3105] [Bug reports] [answered]
Hi, 

I don't see this discussion before.

It has been added Captcha in carelink and it's not possible to bring data to xdrip from there.

Thanks in advance!

**Comment by iPlay-code** [comment]
They added the Captcha 1 or 2 days ago, so no data is brought into Xdrip because of it. 
This message is given in the logs:
"CareLink login error! Response code: 200
Error message: NullPointerException: value == null".

**Comment by RansomTime** [comment]
Looks like this was rolled out to the UK yesterday evening with their downtime. Same login error for me today.

Incredibly frustrating that they're making these changes. My heart goes out to the devs who've got this to work up to now, I hope you can come up with a solution for this too, but it's utterly frustrating that Medtronic doesn't provide us with a good way to get **our data** out of the system easily.

**Reply by wtt604** [reply]
I'm going to phone Medtronic today about this. I don't expect anything to happen, but it would be nice! It's affecting us in Canada too.

**Comment by Navid200** [comment]
Is there a reason you guys keep buying this product if you don't like the official app and you don't like the company constantly changing the protocol to make it hard for alternative apps?

If you keep buying, how would they know you are not happy?

**Reply by kukuchta** [reply]
Yes, there is a good reason. In my place only 3 brands are funded by government today, Medtronic among them. Furthermore if some kid gets hospitalized with new T1D diagnosis, there is usually very narrow choice of pump models that hospital actually bought. In my case, there was Medtronic system or pens. Ah well, you can always take any full paid option😉

**Reply by wtt604** [reply]
This is considerably more polite than the answer that I typed out and deleted.

Thank you.

**Reply by Navid200** [reply]
Have you considered using xDrip in Companion app mode?

**Reply by wtt604** [reply]
No, but I'm looking at it right now, does that allow xdrip to get data from the carelink follow app that I have running on my android 13 phone

**Reply by kukuchta** [reply]
Im using it as a fall-back option always when Carelink is down. But it is not so good for everyday control. There is no info about active insulin, no boluses, no meals, no notifications and no Carelink alerts. Plain BGs. 

And neighbouring BGs often get duplicated, because there is slight mismatch between the time when new value is displayed in minimed Mobile notification and the time when CompanionApp follower reads it:

![Screenshot_2023-09-29-21-20-48-16_2dd2c1bcb8f659cd696bcba2072f751b.jpg](https://github.com/NightscoutFoundation/xDrip/assets/117176531/306d2c72-22de-41ae-825f-13af4f450580)

Its dangerous when there are rapid changes.

**Reply by wtt604** [reply]
I primarily use it to send data to my fitbit, through the glance watchface, so I'd probably be happy if it just gets the companion data, but I think I'm having issues with getting it set up. So far I'm at 1800 minutes since the last signal using the instructions on https://navid200.github.io/xDrip/docs/Follow/CompanionApp.html

**Reply by kukuchta** [reply]
Try here, might be issue with app permissions.

https://xdrip.readthedocs.io/en/latest/install/companion/

**Reply by wtt604** [reply]
This isn't working for me, I'll try again later

**Reply by skylercole** [reply]
Thank you so much. I wasn't even aware of companion app mode. This is def better than nothing.  I use it to get data to my watch as well.

**Reply by jamorham** [reply]
@Navid200 many people do not have a choice of which devices they get to use. Often it can be limited by their health service or insurance. This is not a typical market driven by consumer pressures. We should focus on helping people to gain access to their data and improving their daily lives.

**Reply by Navid200** [reply]
I asked a question to find out if they absolutely had no option or if it was for any other reason.  I focus on trying to improve people's quality of life.  That is why I put together a guide to help people use xDrip.  Why else would I have done that?

The manufacturer clearly doesn't  want us to do what we are doing considering how many times we have found a way and they have switched it on us.  
If you are happy with becoming known as a rogue app, I am OK.  Let's change it again.  And, how long do you think it will take for the manufacturer to add yet another obstacle?  

And is that going to improve their quality of life?  If next month, they are back here asking for finding a way to fix the next hurdle?

**Reply by RansomTime** [reply]
> Is there a reason you guys keep buying this product if you don't like the official app and you don't like the company constantly changing the protocol to make it hard for alternative apps?

I'm an NHS patient who was given the choice of 2 pumps, and was strongly recommended not to go with one due to it not meeting my needs.

> The manufacturer clearly doesn't want us to do what we are doing considering how many times we have found a way and they have switched it on us.

Frankly, I don't care what the manufacturer does or doesn't want me to do with data they hold about my body. It's not their data, it's mine. In a perfect world we wouldn't need to use the Carelink app at all as it will just communicate with xDrip and the Pump directly, but I can imagine that protocol will be very hard to reverse-engineer.

> If you are happy with becoming known as a rogue app, I am OK. Let's change it again. And, how long do you think it will take for the manufacturer to add yet another obstacle?

I guess we'll find out.

**Reply by Navid200** [reply]
I have diabetes myself.  My original question was never meant to suggest that diabetes is not a challenge, or that I don't think we should help you.

We need to have open conversations here, without making it personal.
If my question sounded like I was making a personal attack, I am sorry.

xDrip is used by many.  We need to consider the best approach taking into account the impact our actions may have on all, not a few.
I am sorry if that is a cold statement.  There are many players.  We cannot survive without the manufacturers staying in and investing and doing research and development.  We need to keep the manufacturers happy too.  It is a thin line to tread.

**Reply by RansomTime** [reply]
Water under the bridge, sorry if I was hostile too there.

It's my experience, at least from Abbott and Medtronic that they aren't (and have never been) on our side with this.

Your experience may well differ, however.

I'd like to once again thank you for mentioning the companion mode, it wasn't a feature I was aware of and at least provides the BG values.

**Reply by kukuchta** [reply]
If keeping xDrip application clean in legal terms has become such a concert let's move to the model where application only gives users possibility to include different collection services as external  pluggins. 

This way they could be excluded from the main repository and be developed independently as you call it "rogue" modules. From what I saw in the code, they are already pretty independant entities.

**Reply by wtt604** [reply]
> Try here, might be issue with app permissions.
> 
> https://xdrip.readthedocs.io/en/latest/install/companion/

I still have "BG reading unavailable" on xdrip, and I am using carelink connect to follow my son.

I assume companion mode works in his form, I am not the one with a local CGM that I need xdrip to connect to. I need xdrip to talk to the carelink connect app, that currently gives me up to date information

**Comment by RansomTime** [comment]
Thanks for mentioning companion, that at least lets me get my data on xdrip.

**Comment by palmarci** [comment]
Im researching the topic at https://github.com/benceszasz/xDripCareLinkFollower/issues/37

**Reply by lmbigdata** [reply]
Thanks!!!

**Comment by benceszasz** [comment]
I have released a first working prototype of my private CareLinkFollower for xdrip app for reCAPTCHA. If it is working well, I will create a new pull request for the official xdrip. But first there are few things to do, for example to use reCAPTCHA browser login only for servers outside US, since in the US the reCAPTCHA is not required (yet?).

**Reply by Blacktails111** [reply]
Thank you so much...

Variant 1 worked for me 🙏

**Reply by kukuchta** [reply]
Same here, country - Poland. Many thanks, job well done! 😁

**Reply by robmaggidev** [reply]
Same here in Italy, variant 1 worked, great job! Thanks!

**Reply by awbunge** [reply]
Excuse me, robmaggiedev, but how did you find and get the "Variant 1" you mention to work? Right now I find myself really lost, and bewildered, not knowing what to do....

**Reply by awbunge** [reply]
Dear benceszasz, thank you very much. You did an excellent job. "Variant 1" works very good, in Rome, Italy. I guess an official version is needed to make it work with the Garmin data fields offered by xDrip+ via Nightscout (e.g. NightscoutData2 NSS xDrip+ CGM data field). Anyway, we (all, I think) hope you can offer it as soon as possible. Again, thank you very much.

**Reply by Ghubcd** [reply]
Same here, country - Lithuania. Thanks!!!

**Reply by lmbigdata** [reply]
I tried to kill the app for several hours and I obtained token expired. Then I logged in again and it's working perfectly again.

Is this the expected behaviour @benceszasz ?

**Comment by robmaggidev** [comment]
I clicked benceszasz github profile page, then I went to the latest release
of the carelink follower project itself, here
https://github.com/benceszasz/xDripCareLinkFollower/releases/tag/v0.1.13-beta
and downloaded the apk of variant 1 (from the assets). If I understand
correctly this has been made just as a beta version to check if it works,
so the code will be merged to the official xdrip build.

Il Lun 2 Ott 2023, 15:05 awbunge ***@***.***> ha scritto:

**Reply by awbunge** [reply]
Thank you very much, dear robmaggidev. I did as you said, I installed this "Variant 1", and it works very well.
Country: Italy

**Comment by jhautot** [comment]
Variant 1 is working well on my side as well. Hope you can merge as soon as possible.
Thanks a lot !
PS: Country Belgium

**Comment by awbunge** [comment]
Since this morning, "Variant 1" no longer works. The most frequent error is "Error renewing token". I also got the error "Aggressively restarting collector service due to lack of reception: backoff: 300". I did again, in manual mode, the "login", and everything worked normally again (I am starting to get seriously angry with Medtronic).
![WhatsApp Image 2023-10-04 at 10 16 54](https://github.com/NightscoutFoundation/xDrip/assets/146633543/73e770a8-e1b4-4eda-959d-7a81b59bbce5)

**Reply by lmbigdata** [reply]
Maybe you lost internet connection and it was not possible to renew token. It's happening to me when loosing internet connection

**Reply by robmaggidev** [reply]
I'm not sure and not an expert, but I think this is how it is supposed to work since Medtronic introduced ReCaptcha in the login process. I mean: your login session is time-limited and expires in X hours (don't know how many hours, but it's not limitless), therefore you need to re-login from time to time.

**Reply by awbunge** [reply]
Thank you, Big Man, you gave me a Big tip.

**Reply by awbunge** [reply]
Thank you also, robmaggidev.

**Reply by wtt604** [reply]
I see a lot of talk about variant 1, I installed the base variant on Sunday morning PST, and it's been rock solid since then, on a Google pixel.

Is there any reason that variant 1 seems to be the most popular?

**Reply by jhautot** [reply]
I think the only reason is that it is the first package. Everyone tried the first link and as it was working so no need to try others.

**Reply by Navid200** [reply]
@wtt604 You can install a variant while still having the main xDrip installed.  You will not be able to install the main version if you also have the official main version installed.

A variant has a different database with respect to the main xDrip.
So, if your intention is to maintain an intact database containing all your readings as you switch from the official xDrip to this variant, you are mistaken and your database is not going to contain all your readings.

**Reply by wtt604** [reply]
@Navid200 ahh, thanks, no, all I care about is just getting the data to my Fitbit, so I'm very happy with it, I'm not that bothered about databases being continued, or historical data.

I'm happy it does all that, but I'm not needing that functionality.

Thanks again to all involved

**Reply by robmaggidev** [reply]
@Navid200 so... if I install the main build (not one of the variants) I'll be already able to send my data to Nightscout wodgets on my Garmin device?

**Reply by Navid200** [reply]
@robmaggidev I'm not sure I understand your question.

You cannot install multiple versions of xDrip on the same phone.
Years ago, parents who had multiple diabetic children asked for help because they didn't want to carry multiple phones.
The option offered to them was to use xDrip and a variant, which can be installed on the same phone at the same time.  They are identical except the name.   Use each for one kid.  That was the purpose of releasing variants, having multiple sources.

But, if you use a variant for yourself, you need to know that was never the intent.  You can.  But, you need to understand the consequences.
So, I just wanted to help clarify.

**Reply by wtt604** [reply]
> @Navid200 so... if I install the main build (not one of the variants) I'll be already able to send my data to Nightscout wodgets on my Garmin device?

For me, I wasn't able to get the glance watchface to talk to the variant, I think it shows up as a different program, so maybe your Garmin device has the same issue, but the main build has been rock solid for me since Sunday morning.

**Reply by robmaggidev** [reply]
thanks @Navid200 and @wtt604, what I was asking actually was: there are several apk of this build https://github.com/benceszasz/xDripCareLinkFollower/releases/tag/v0.1.13-beta from @benceszasz, so if I install the one with no version (so this one [xdrip-carelink.apk](https://github.com/benceszasz/xDripCareLinkFollower/releases/download/v0.1.13-beta/xdrip-carelink.apk)) am I able to see my data on my Garmin device now as I'm able to do it with the official xDrip apk without waiting for the ReCaptcha fix to be stable and merged to the official branch? And the answer is yes :) because I tried and it works.

Of course I'll use it at my own risk, I know it's a beta and could be unstable, but I need it just for myself: I use xDrip to have glicemy level on a Garmin cyclo-computer while I go out with my gravel bike, and it's far too handy than to check it from the phone or from the micropump. If after some time it stops working, no problem for me, I'll check the glicemy from the micro.

**Reply by OGianni** [reply]
thanks Roberto, variant 1 was very stable for me, but I just lacked the connection with Garmin for sports use. Thanks to everyone for the enormous effort and to Roberto for the icing.I'm not an expert on this platform, do you have a watchface to recommend, I don't want to disturb here, can we talk in private?

**Reply by awbunge** [reply]
@robmaggidev, I did it, and with the xdrip-carelink.apk my clock Garmin Fenix 7 (that I use specially for running)  started working perfectly again (with Nightscout [NightscoutData2 NSS xDrip+ CGM data field]).

**Comment by awbunge** [comment]
Dear friends, two days ago appeared on this site, https://github.com/NightscoutFoundation/xDrip/releases, a new version, "Nightly build 6th Oct 2023". For a minute I had the illusion that it might work, overcoming the "captcha" barrier set by Medtronic since the last week of September. Unfortunately, I realized that it was not. That version from October 6 at night does not allow access. unlike the one we are discussing here. Can anyone tell me what is really going on? This seems to be a never ending battle, and pointless...

**Reply by benceszasz** [reply]
I have released beta versions of the CareLink Follower for reCAPTCHA in my private repository:
https://github.com/benceszasz/xDripCareLinkFollower
The latest version was this one:
https://github.com/benceszasz/xDripCareLinkFollower/releases/tag/v0.1.14-beta
This private version can be used temporary until the new CareLink Follower is finalized and merged into the official xDrip. Today I have created a pull request of CareLink Follower with reCAPTCHA support for the official xDrip, because it seems to be stable enough to get into the main version: https://github.com/NightscoutFoundation/xDrip/pull/3121
After this pull request is approved, the next official releases will contain it.

**Reply by awbunge** [reply]
I already tested this new version v0.1.14-beta (in my Smartphone Samsung Galaxy S21, Android 13). It works very well. Congratulations, and thank you very much.

**Reply by cris-mo** [reply]
Hello everyone, I really have to log in to xdrip every day with the reCAPTCHA.

How can it be possible that I stay logged in?

**Reply by skritland** [reply]
For me, I have to solve reCaptcha much less often when I use the carelink follower account only in the xDrip.

**Comment by benceszasz** [comment, marked as answer]
CareLink Follower with reCAPTCHA using browser login has been merged into the official xDrip version yesterday (08/10/2023). Latest official xDrip release already contains this new version of CareLink Follower with reCAPTCHA using browser login:
https://github.com/NightscoutFoundation/xDrip/releases/tag/2023.10.08
Please try to use a dedicated CareLink Care Partner account for xDrip to avoid closing your xDrip session by logging in with the same account from either the CareLink Personal website or a Medtronic offical app.

**Reply by lmbigdata** [reply]
Thanks! Already installed!

**Reply by PhilTU** [reply]
on Facebook a user reckoned the nightly didn't work 
![387176733_10160042202822695_7199390903870934728_n](https://github.com/NightscoutFoundation/xDrip/assets/76240903/af356dec-bcf7-48b6-9ccf-391a88669e74)

I asked them to feedback here. The private version did work, apparently.

**Reply by Navid200** [reply]
@PhilTU If you read Bence's comments in the pull request (https://github.com/NightscoutFoundation/xDrip/pull/3121), you will see that lots of testing was done and everything was merged into the official repository.
He has done a lot of work already.

Since then, many have reported that the official release is working for them.

You are saying that someone has told you that it isn't working for them.  Why can they not present their case themselves?
What do you expect Bence to do now seeing your comment?
What would you do if you were in Bence's shoes now?

Would you please ask the person who has reported this to you to open a new discussion themselves?
It's not going to work if you become a go-between.  We need the person who is experiencing the problem to come here and report it.

**Reply by PhilTU** [reply]
Thankyou Navid, I appreciate that many have it working and that much good work was done addressing this issue when it arose. I have already asked the person concerned to contribute here and provided a link for them, I have no interest in being an intermediary.

If I were in a position of knowledge I would look at the screenshot and diagnose that the user may be doing something wrong, or look at the URL and see if it gave me a steer why that person has an issue. But I am not, so I merely flagged it here for _discussion_ 

 It may, perhaps, be a geographical thing as the server in the screenshot doesn't respond to a web browser.
 
 But let's shoot the messenger, eh ?

**Reply by Navid200** [reply]
I didn't, and won't, shoot the messenger.
Since I know you contribute significantly in the facebook group, and I want your time to be used efficiently, as we don't have enough people like you, I suggested to you to consider that there are so many factors that could contribute to what the individual experiences.  We have no idea what version he is running now.  We only have his word for it.  
Bence is going to ask the same question.  Looking at that screenshot, there is no way to know which version he is using.

If enough people use the official release and follow all your instructions and they all find it is not working, we will certainly have a problem and we can open a discussions here.

**Reply by lmbigdata** [reply]
@PhilTU This was happening to me when I don't have internet connnection

**Reply by robmaggidev** [reply]
Sorry guys, I downloaded that apk (20231008) but I can't see the Carlink Follower option during the configuration process, what am I missing? It has bene moved somewhere else? Thanks

![Screenshot_20231010_173641_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/12560353/b1dfea07-a0b5-424a-94f8-73c3cf39f525)

**Reply by awbunge** [reply]
Exactly. "Net:: ERR_TIMED_OUT" means a difficulty with the de Internet connection.

**Reply by wtt604** [reply]
@robmaggidev it's not in there, but when you go into the menu on the left side if the screen, it is there. I had the same issue

**Reply by Navid200** [reply]
@robmaggidev Please go to the hardware data source page to see the option: https://navid200.github.io/xDrip/docs/HardwareDataSource.html

**Comment by awbunge** [comment]
Great! It works well. Thank you very much, @[benceszasz...

**Comment by robmaggidev** [comment]
Thanks, perfect, I found it!

Il Mar 10 Ott 2023, 17:56 Navid ***@***.***> ha scritto:

**Comment by PhilTU** [comment]
> Exactly. "Net:: ERR_TIMED_OUT" means a difficulty with the de Internet connection.

or the server, I got no response from the server with my functioning internet connection.

**Comment by PhilTU** [comment]
> Sorry guys, I downloaded that apk (20231008) but I can't see the Carlink Follower option during the configuration process, what am I missing? It has bene moved somewhere else? 

that's a wizard, for those who believe in magic. Dismiss it and go the Settings in the app.
